### PR TITLE
Feature/peer service

### PR DESCRIPTION
**Pull request by satellitex**
Create Peer-Service

 - addPeer( peer::Node& )
 - removePeer( peer::Node& )
 - updatePeer( peer::Node& )
 - isLeaderMyPeer( peer::Node & )

Use "publicKey" for judging the same peer.

How to call :: peer_service_with_json_test.cpp

### CODE REVIEW

**Comment by Warchant** [comment]
I will fix conflicts

**Comment by Warchant** [comment]
Conflict is resolved, branch can be merged

File: core/consensus/sumeragi.cpp [hunk 1]
```diff
@@ -170,18 +171,50 @@ namespace sumeragi {
...
+        {
+            auto peers = config::PeerServiceConfig::getInstance().getPeerList();
+            for (auto&& p : peers ) {
+                validatingPeers.push_back( std::move(p) );
+            }
+
+            this->numValidatingPeers = this->validatingPeers.size();
+            // maxFaulty = Default to approx. 1/3 of the network.
+            this->maxFaulty = config::IrohaConfigManager::getInstance()
+                    .getMaxFaultyPeers(this->numValidatingPeers / 3);
```
Comment (takemiyamakoto) [inline core/consensus/sumeragi.cpp]: Should be: ```(this->numValidatingPeers-1)/3```. Also, there should be a check to set this to "0" if the result is less than 0.

File: core/consensus/sumeragi.cpp [hunk 2]
```diff
@@ -170,18 +171,50 @@ namespace sumeragi {
...
+            }
+
+            if (this->proxyTailNdx >= this->validatingPeers.size()) {
+                this->proxyTailNdx = this->validatingPeers.size() - 1;
+            }
+
+            this->panicCount = 0;
+            this->myPublicKey = config::PeerServiceConfig::getInstance().getMyPublicKey();
+
+            this->isSumeragi = this->validatingPeers.at(0)->getPublicKey() == this->myPublicKey;
```
Comment (takemiyamakoto) [inline core/consensus/sumeragi.cpp]: There could be potential race conditions. Probably best not to set the isSumeragi here, but rather get this information from the peer service when starting to process.

File: core/infra/config/peer_service_with_json.cpp [hunk 3]
```diff
@@ -56,26 +78,215 @@ std::string PeerServiceConfig::getMyIp() {
...
+        it->setTrustScore(peer.getTrustScore());
+    }
+
+  } catch ( exception::service::UnExistFindPeerException& e ) {
+    logger::warning("updatePeer") << e.what();
+    return false;
+  } catch( exception::service::DuplicationPublicKeyException& e ) {
+      logger::warning("updatePeer") << e.what();
+  } catch ( exception::service::DuplicationIPException& e ) {
+    logger::warning("udpatePeer") << e.what();
```
Comment (takemiyamakoto) [inline core/infra/config/peer_service_with_json.cpp]: udpatePeer→updatePeer

File: core/consensus/sumeragi.cpp (hunk 1 above)
Comment (satellitex) [inline core/consensus/sumeragi.cpp]: change
 `(this->numValidatingPeers-1)/3`
to
`std::max( 0, this->numValidatingPeers / 3 )`

File: core/consensus/sumeragi.cpp (hunk 2 above)
Comment (satellitex) [inline core/consensus/sumeragi.cpp]: I just thinking this problem now.
I want to talk to you.

File: core/infra/config/peer_service_with_json.cpp (hunk 3 above)
Comment (satellitex) [inline core/infra/config/peer_service_with_json.cpp]: ok

**Pull request merged**